ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java

```diff
@@ -454,394 +428,362 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
-      e.printStackTrace();
-      System.out.println("Unknown exception thrown " + e.getMessage());
+    } finally {
+      if (stream != null) {
+        stream.close();
+      }
     }
-    return false;
   }
 
-  public synchronized void logOutput(String msg) {
-    try {
-      if (outFile == null) {
-        System.out.println(msg);
+  /**
+   * Compares every entry in the log to those in the index. Checks to see if the status of the blob is consistent with
+   * each other
+   * @param logFile the log file to be dumped
+   * @param blobList List of BlobIds to be filtered for, Can be {@code null}
+   * @param replicaRootDirectory the root directory of the replica
+   * @param generateBlobStatusReport {@code true} if verbose report about status of each blob needs to be logged.
+   *        {@code false} otherwise
+   * @param filter {@code true} if needs to be filtered, {@code false} otherwise
+   * @throws Exception
+   */
+  public void compareLogEntriestoIndex(String logFile, ArrayList<String> blobList, String replicaRootDirectory,
+      boolean filter, boolean generateBlobStatusReport)
+      throws Exception {
+    if (logFile == null || replicaRootDirectory == null) {
+      log("logFile and replicaRootDirectory needs to be set for compareLogToIndex", LogLevel.INFO);
+      System.exit(0);
+    }
+    ConcurrentHashMap<String, BlobStatus> blobIdToBlobStatusMap =
+        dumpIndexesForReplica(replicaRootDirectory, blobList, generateBlobStatusReport);
+    ConcurrentHashMap<String, DumpDataHelper.LogBlobRecord> blobIdToLogRecordStats = new ConcurrentHashMap<>();
+    dumpDataHelper.dumpLog(new File(logFile), 0, -1, blobList, filter, blobIdToLogRecordStats);
+    long totalInconsistentBlobs = 0;
+
+    for (String blobId : blobIdToLogRecordStats.keySet()) {
+      DumpDataHelper.LogBlobRecord logBlobRecord = blobIdToLogRecordStats.get(blobId);
+      if (blobIdToBlobStatusMap.containsKey(blobId)) {
+        BlobStatus blobStatus = blobIdToBlobStatusMap.get(blobId);
+        if ((logBlobRecord.isDeleted || logBlobRecord.isExpired) && !blobStatus.isDeletedOrExpired) {
+          log("Blob " + blobId + " is deleted/expired in log while alive in index", LogLevel.INFO);
+          totalInconsistentBlobs++;
+        }
       } else {
-        fileWriter.write(msg + "\n");
+        if (!(logBlobRecord.isConsistent && (logBlobRecord.isDeleted || logBlobRecord.isExpired))) {
+          totalInconsistentBlobs++;
+          log("Blob " + blobId + " found in Log but not in index", LogLevel.INFO);
+          log("Log Record details : " + logBlobRecord, LogLevel.INFO);
+        }
       }
-    } catch (IOException e) {
-      System.out.println("IOException while trying to write to File");
     }
+    log("Total Inconsistent blobs count " + totalInconsistentBlobs, LogLevel.INFO);
   }
 
-  public void shutdown() {
-    try {
-      if (outFile != null) {
-        fileWriter.flush();
-        fileWriter.close();
+  class MergedIntervals {
```

Comment:
Doc. And can you make more explanation how this class internally works? How do you check if it is fully covered? This class does not provide a method to do so.